### ISSUE TITLE
text token index slice to N-1

### ISSUE THREAD

**Issue by j-min** [kakaobrain/minDALL-E#16]
Hi, thanks for sharing the code.

In the [`forward` function](https://github.com/kakaobrain/minDALL-E/blob/main/dalle/models/stage2/transformer.py#L89-L90) of Transformer1d,
text index is sliced with `0 ~ N-2` and image index is sliced with `N-1 ~ N-1 + (T-1)`.

```python3
B, T = images.shape
_, N = texts.shape
...
x = torch.cat([texts, images], axis=1).contiguous()
...
texts = x[:, :N-1].contiguous()
images = x[:, N-1:-1].contiguous()
```

Could you please clarify why you didn't slice like below? Thanks!
```python3
texts = x[:, :N]
images = x[:, N:]
```

**Comment by j-min** [comment]
Oh I guess it's for autoregressive next token prediction objective.
Then do you use loss like this? It would be great if you could share your actual loss implementation.
```python3
# [B, N-1]
label_txt = texts[:, 1:]

# [B, T]
label_code = images

loss_text = F.cross_entropy(logits_txt.view(-1, logits_txt.shape[-1]), label_txt.view(-1))
loss_img = F.cross_entropy(logits_img.view(-1, logits_img.shape[-1]), label_code.view(-1))

loss = loss_text + loss_img 
```

**Comment by LeeDoYup** [comment]
@j-min  Thanks for your interest in our project !
As you guess, the index slicing is for the training of next token prediction.
Please note that we do not concatenate an additional token such as CLS or SOS at the head of texts or images.
Thus, the first code of an image is predicted at the last position of the text tokens.

In addition, as you have mentioned, 
`loss_text` uses `texts[:,1:]`, and `loss_img`uses `images` for the label, respectively.
